quinn-proto/src/connection.rs

```diff
@@ -750,12 +750,32 @@ impl Connection {
         remote: SocketAddrV6,
         partial_decode: PartialDecode,
     ) -> Option<BytesMut> {
-        let crypto = if partial_decode.is_handshake() {
-            &self.handshake_crypto
+        let result = if partial_decode.is_handshake() {
+            partial_decode.finish(self.handshake_crypto.pn_decrypt_key())
+        } else if partial_decode.key_phase() == self.key_phase {
+            partial_decode.finish(self.crypto.as_ref().unwrap().pn_decrypt_key())
         } else {
-            &self.crypto.as_ref().unwrap()
+            let new = self.crypto.as_mut().unwrap().update();
+            let decoded = partial_decode.finish(new.pn_decrypt_key());
+
+            if let Ok((
+                Packet {
+                    header: Header::Short { number, .. },
+                    ..
+                },
+                ..
+            )) = decoded
+            {
+                let decoded_number = number.expand(self.rx_packet + 1);
+                let old = mem::replace(self.crypto.as_mut().unwrap(), new);
+                self.prev_crypto = Some((decoded_number, old));
+                self.key_phase = !self.key_phase;
```

Comment:
The key update must not be committed until we have successfully authenticated the packet's payload. Otherwise, connections can be trivially DoSd. One approach would be to pass an optional `&Crypto` to use in preference to the one from `self` down through `handle_packet` to `decrypt_packet`, but there might be a cleaner approach.